pkg/modifycontroller/modify_volume.go

```diff
@@ -0,0 +1,178 @@
+			targetVacName = pvc.Status.ModifyVolumeStatus.TargetVolumeAttributesClassName
+		}
+		if *curVacName == targetVacName {
+			return ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
+		} else {
+			// Check if the PVC is in uncertain State
+			pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+			if err != nil {
+				return pvc, pv, err, false
+			}
+			_, ok := ctrl.uncertainPVCs[pvcKey]
+			if !ok {
+				// PVC is not in uncertain state
+				klog.V(3).InfoS("previous operation on the PVC failed with a final error, retrying")
+				return ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
+			} else {
+				vacObj, _, err := ctrl.vacInformer.GetStore().GetByKey(*pvcSpecVacName)
+				if err != nil {
+					return pvc, pv, err, false
+				}
+				return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vacObj.(*storagev1alpha1.VolumeAttributesClass), pvcSpecVacName)
+			}
+		}
+
+	}
+	// No modification required
+	return pvc, pv, nil, false
+}
+
+// func validateVACAndModifyVolumeWithTarget validate the VAC. The function sets pvc.Status.ModifyVolumeStatus
+// to Pending if VAC does not exist and proceeds to trigger ModifyVolume if VAC exists
+func (ctrl *modifyController) validateVACAndModifyVolumeWithTarget(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
+	// The controller only triggers ModifyVolume if pvcSpecVacName is not nil nor empty
+	pvcSpecVacName := pvc.Spec.VolumeAttributesClassName
+	// Check if pvcSpecVac is valid and exist
+	vacObj, exists, err := ctrl.vacInformer.GetStore().GetByKey(*pvcSpecVacName)
+	if exists && err == nil {
+		// Mark pvc.Status.ModifyVolumeStatus as in progress
+		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInProgress, nil)
+		if err != nil {
+			return pvc, pv, err, false
+		}
+		// Record an event to indicate that external resizer is modifying this volume.
+		ctrl.eventRecorder.Event(pvc, v1.EventTypeNormal, util.VolumeModify,
+			fmt.Sprintf("external resizer is modifying volume %s with vac %s", pvc.Name, vacObj.(*storagev1alpha1.VolumeAttributesClass).Name))
+		return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vacObj.(*storagev1alpha1.VolumeAttributesClass), pvcSpecVacName)
+	} else {
+		// Mark pvc.Status.ModifyVolumeStatus as pending
+		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumePending, nil)
+		return pvc, pv, err, false
+	}
+}
+
+// func controllerModifyVolumeWithTarget trigger the CSI ControllerModifyVolume API call
+// and handle both success and error scenarios
+func (ctrl *modifyController) controllerModifyVolumeWithTarget(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume,
+	vacObj *storagev1alpha1.VolumeAttributesClass,
+	pvcSpecVacName *string) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
+	var err error
+	pvc, pv, err = ctrl.callModifyVolumeOnPlugin(pvc, pv, vacObj)
+	if err == nil {
```

Comment:
There is some duplication between `callModifyVolumeOnPlugin` function and code here. 

Inside `callModifyVolumeOnPlugin` function:

If there is an error and error is a final error, then we are marking volume as infeasible.  Inside `controllerModifyVolumeWithTarget` function - if returned error is `InvalidArgument`, then we are marking volume as infeasible. So which is it?  The former is a superset of later - isn't it?


IMO, this should be cleaned up.

Reply:
Moving all the error handling in controllerModifyVolumeWithTarget instead

Reply:
So, this is what I don't understand. Both `InvalidArgument` errors and any other error that is a final error causes `PersistentVolumeClaimModifyVolumeInfeasible` status. So - why check specifically for `InvalidArgument`? It seems to me that just checking for `util.IsFinalError(err)` should be enough for setting `Infeasible` state.

Reply:
Invalid argument is not a final error, IsFinalError does not include invalid argument?

```
case codes.Canceled, // gRPC: Client Application cancelled the request
		codes.DeadlineExceeded,  // gRPC: Timeout
		codes.Unavailable,       // gRPC: Server shutting down, TCP connection broken - previous volume operation may be still in progress.
		codes.ResourceExhausted, // gRPC: Server temporarily out of resources - previous volume operation may be still in progress.
		codes.Aborted:           // CSI: Operation pending for volume
		return false
	}
```

Reply:
We do not removePVCFromModifyVolumeUncertainCache if it is InvalidArgument

Reply:
> We do not removePVCFromModifyVolumeUncertainCache if it is InvalidArgument, for other final errors we do.

Why? If we are treating all final errors as infeasible, what is special about `InvalidArgument` error? cc @msau42

Reply:
Clarified with Michelle and updated here: https://github.com/kubernetes-csi/external-resizer/pull/351/commits/f516a04360f108116b08aeb1744e6743387b3cd5

Only invalidarugment can transition to infeasible so the users know they need to change something(params etc.).

Reply:
`InvalidArgument` should be the only error that results in `Infeasible` status. `InvalidArgument` is a type of final error, so we should still treat it as such.

Reply:
Okay yeah that makes sense, because that is what we agreed on during design. The implementation however still appears to be incorrect.  `InvalidArgument` error is also a final error @sunnylovestiramisu

Reply:
Yeah but we treat them differently:

if status.Code() == codes.InvalidArgument {
					// Mark pvc.Status.ModifyVolumeStatus as infeasible
					pvc, markModifyVolumeInfeasibleError := ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInfeasible, err)
					if markModifyVolumeInfeasibleError != nil {
						return pvc, pv, markModifyVolumeInfeasibleError, false
					}
				}


And then for all the final errors including invalid argument, we remove them from the uncertain cache.

Reply:
oh okay. cool. This looks good now. thank you.